src/compressed_tensors/compressors/model_compressors/model_compressor.py

```diff
@@ -301,23 +308,41 @@ def decompress(self, model_path: str, model: Module):
         :param model: pytorch model to load decompressed weights into
         """
         model_path = get_safetensors_folder(model_path)
+        sparse_decompressed = False
+
         if self.sparsity_compressor is not None:
+            # Sparse decompression is applied on the model_path
             dense_gen = self.sparsity_compressor.decompress(model_path)
             self._replace_weights(dense_gen, model)
             setattr(model, SPARSITY_CONFIG_NAME, self.sparsity_compressor.config)
+            sparse_decompressed = True
 
         if self.quantization_compressor is not None:
-            names_to_scheme = apply_quantization_config(model, self.quantization_config)
-            load_pretrained_quantization(model, model_path)
+            # Temporarily set quantization status to FROZEN to prevent
+            # quantization during apply_quantization_config. This ensures
+            # that the dtypes of the weights are not unintentionally updated.
+            # The status is restored after quantization params are loaded.
+            with override_quantization_status(
+                self.quantization_config, QuantizationStatus.FROZEN
+            ):
+                names_to_scheme = apply_quantization_config(
+                    model, self.quantization_config
+                )
+                load_pretrained_quantization(model, model_path)
+
+            model_path_or_state_dict = (
```

Comment:
Why is this special logic necessary? Haven't the model weights already been replaced by _replace_weights in the sparse case?

In both cases, it seems that we can reduce the amount of special logic by just loading the model state dict?

Reply:
Because it is not necessary that we will always have a sparse_compressor, think about cases with just quantization, in those cases quantization compression would have to operate on the model path. We can for sure improve it like separating the loading step from the actual compression/decompression but we can do that in a follow up PR

Reply:
> We can for sure improve it like separating the loading step from the actual compression/decompression but we can do that in a follow up PR

Yeah this is what I was referring to

Reply:
That's fine